tools/wptrunner/wptrunner/executors/executorwebdriver.py

```diff
@@ -0,0 +1,375 @@
+                   strip_server)
+from .protocol import (BaseProtocolPart,
+                       TestharnessProtocolPart,
+                       Protocol,
+                       SelectorProtocolPart,
+                       ClickProtocolPart,
+                       SendKeysProtocolPart,
+                       TestDriverProtocolPart)
+from ..testrunner import Stop
+
+import webdriver as client
+
+here = os.path.join(os.path.split(__file__)[0])
+
+webdriver = None
+
+class WebDriverBaseProtocolPart(BaseProtocolPart):
+    def setup(self):
+        self.webdriver = self.parent.webdriver
+
+    def execute_script(self, script, async=False):
+        method = self.webdriver.execute_async_script if async else self.webdriver.execute_script
+        return method(script)
+
+    def set_timeout(self, timeout):
+        try:
+            self.webdriver.timeouts.script = timeout
+        except client.WebDriverException:
+            # workaround https://bugs.chromium.org/p/chromedriver/issues/detail?id=2057
+            body = {"type": "script", "ms": timeout * 1000}
+            self.webdriver.send_session_command("POST", "timeouts", body)
+
+    @property
+    def current_window(self):
+        return self.webdriver.window_handle
+
+    def set_window(self, handle):
+        self.webdriver.window_handle = handle
+
+    def wait(self):
+        while True:
+            try:
+                self.webdriver.execute_async_script("")
+            except client.TimeoutException:
+                pass
+            except (socket.timeout, client.NoSuchWindowException,
+                    client.UnknownErrorException, IOError):
+                break
+            except Exception as e:
+                self.logger.error(traceback.format_exc(e))
+                break
+
+
+class WebDriverTestharnessProtocolPart(TestharnessProtocolPart):
+    def setup(self):
+        self.webdriver = self.parent.webdriver
+
+    def load_runner(self, url_protocol):
+        url = urlparse.urljoin(self.parent.executor.server_url(url_protocol),
+                               "/testharness_runner.html")
+        self.logger.debug("Loading %s" % url)
+
+        self.webdriver.url = url
+        self.webdriver.execute_script("document.title = '%s'" %
+                                      threading.current_thread().name.replace("'", '"'))
+
+    def close_old_windows(self):
+        exclude = self.webdriver.window_handle
+        handles = [item for item in self.webdriver.handles if item != exclude]
+        for handle in handles:
+            try:
+                self.webdriver.window_handle = handle
+                self.webdriver.close()
+            except client.NoSuchWindowException:
+                pass
+        self.webdriver.window_handle = exclude
+        return exclude
+
+    def get_test_window(self, window_id, parent):
+        test_window = None
+        if window_id:
+            try:
+                # Try this, it's in Level 1 but nothing supports it yet
```

Comment:
Level 1 of what? https://w3c.github.io/webdriver/#dfn-execute-script is presumably widely supported, and WebDriver doesn't have level, so presumably not that.

Reply:
https://w3c.github.io/webdriver/#dfn-json-serialization-of-the-windowproxy-object (WebDriver does have levels)

Reply:
Ah, pointing at that would be nice.